### PR TITLE
[Sofa.Config][MSVC] Change SIMD cmake option and flags

### PR DESCRIPTION
**Pull request by fredroy**
(Only MSVC)

 - SOFA_VECTORIZE is supposed to enable SSE2 but x64 config in MSVC is already enabling ss2 by default. And supposedly now people only use x64. Therefore, if SOFA_VECTORIZE is enabled, msvc is warning (for each cpp) that the sse2 flag is useless.
- SOFA_VECTORIZE also enables fast_math in the same time (maybe you would not want)
- the name itself is not really informative (vectorize what ?)

This PR:
 - add 2 new options: SOFA_ENABLE_FAST_MATH (and enable fast_math); and SOFA_ENABLE_SIMD, which will enable either AVX2 or AVX or nothing if the CPU supports respectively either AVX2 or AVX or nothing (SSE2 is the fallback in msvc/x64)
 - deprecates SOFA_VECTORIZE , which enables the 2 previous options if checked.

More lecture for the curious: https://devblogs.microsoft.com/cppblog/do-you-prefer-fast-or-precise/
fpfast enables auto-vectorisation (among other stuff), thats why they are usually enabled together.

On my setup, with SOFA_ENABLE_FAST_MATH + SOFA_ENABLE_SIMD, I can get around 0 to 10% of speedup. Nothing revolutionary anyway. 
TorusFall.scn (1000steps):
 - fastfp+avx2: `7.90001 s ( 126.582 FPS)`
 - fastfp `8.3388 s ( 119.921 FPS)`
 - avx2  `8.49847 s ( 117.668 FPS)`
 - ss2 `8.56825 s ( 116.71 FPS)`
We can see that it should be either both of them enabled, or not. That's why it will not be enabled by default, except if we want fp:fast by default

SIde note, seems Eigen3 is quite optimized with SIMD (w/o vs w/):
- `BM_Matrix_eigenmat33_invert/512`  :  4.45 us   vs   3.40 us
- `BM_Matrix_eigenmat33_determinant/512` : 1.09 us vs  0.882 us 
- `BM_Matrix_eigenmatf_vecmult<24>/512` : 98.0 us  vs 76.0 us

In a separate PR, we could do the same for gcc/clang (with `march=native` and fast_math as well)

EDIT: `SOFA_ENABLE_SIMD` was originally `DSOFA_ENABLE_AVX` but change the name for the future PR for gcc/clang
______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by alxbilger** [comment]
What about AVX512?

**Comment by fredroy** [comment]
> What about AVX512?

I tried activating `/arch:AVX512` but it makes runSofa crash 🥴 so I did not investigate further.
And apparently AVX512 is quite special and not really recommended here and there on the Internets: https://www.phoronix.com/scan.php?page=news_item&px=Linus-Torvalds-On-AVX-512 (could be related to the crash?)

**Comment by fredroy** [comment]
The windows_vs2017_option is built with the 2 options enabled (`-DSOFA_ENABLE_FAST_MATH=ON -DSOFA_ENABLE_AVX=ON`)

**Comment by fredroy** [comment]
Seems VS2017 is not happy 
```
j:\0\src\sofakernel\modules\sofabasemechanics\src\sofabasemechanics\diagonalmass.inl(573) : fatal error C1001: An internal error has occurred in the compiler.
(compiler file 'd:\agent\_work\1\s\src\vctools\compiler\utc\src\p2\main.c', line 187)
 To work around this problem, try simplifying or changing the program near the locations listed above.
Please choose the Technical Support command on the Visual C++ 
 Help menu, or open the Technical Support help file for more information
 ```
 but we are ditching vs2017, right ? 😗

Or it could be that the CI is on a VM, maybe the vCPU is also at fault here.

**Comment by fredroy** [comment]
Interesting, I compiled on the CI with just the FAST_MATH option enabled and those tests fails
```
[UnitTests.SofaGeneralExplicitOdeSolver_test/RungeKutta2ExplicitSolverDynamic_test.rungeKutta2ExplicitSolverDynamicTest_high_dt/0](https://ci.inria.fr/sofa-ci-dev/job/sofa-framework/job/PR-2652/CI_CONFIG=windows_vs2017,CI_PLUGINS=options,CI_TYPE=release/3/testReport/junit/UnitTests/SofaGeneralExplicitOdeSolver_test_RungeKutta2ExplicitSolverDynamic_test/rungeKutta2ExplicitSolverDynamicTest_high_dt_0/)
[UnitTests.SofaGeneralExplicitOdeSolver_test/RungeKutta4ExplicitSolverDynamic_test.rungeKutta4ExplicitSolverDynamicTest_medium_dt/0](https://ci.inria.fr/sofa-ci-dev/job/sofa-framework/job/PR-2652/CI_CONFIG=windows_vs2017,CI_PLUGINS=options,CI_TYPE=release/3/testReport/junit/UnitTests/SofaGeneralExplicitOdeSolver_test_RungeKutta4ExplicitSolverDynamic_test/rungeKutta4ExplicitSolverDynamicTest_medium_dt_0/)
[UnitTests.SofaGeneralExplicitOdeSolver_test/RungeKutta4ExplicitSolverDynamic_test.rungeKutta4ExplicitSolverDynamicTest_small_dt/0](https://ci.inria.fr/sofa-ci-dev/job/sofa-framework/job/PR-2652/CI_CONFIG=windows_vs2017,CI_PLUGINS=options,CI_TYPE=release/3/testReport/junit/UnitTests/SofaGeneralExplicitOdeSolver_test_RungeKutta4ExplicitSolverDynamic_test/rungeKutta4ExplicitSolverDynamicTest_small_dt_0/)
[UnitTests.SofaGeneralImplicitOdeSolver_test/VariationalSymplecticExplicitSolverDynamic_test.variationalSymplecticExplicitSolverDynamicTest_small_dt_without_damping/0](https://ci.inria.fr/sofa-ci-dev/job/sofa-framework/job/PR-2652/CI_CONFIG=windows_vs2017,CI_PLUGINS=options,CI_TYPE=release/3/testReport/junit/UnitTests/SofaGeneralImplicitOdeSolver_test_VariationalSymplecticExplicitSolverDynamic_test/variationalSymplecticExplicitSolverDynamicTest_small_dt_without_damping_0/)
[UnitTests.SofaGeneralImplicitOdeSolver_test/VariationalSymplecticExplicitSolverDynamic_test.variationalSymplecticExplicitSolverDynamicTest_medium_dt_with_damping/0](https://ci.inria.fr/sofa-ci-dev/job/sofa-framework/job/PR-2652/CI_CONFIG=windows_vs2017,CI_PLUGINS=options,CI_TYPE=release/3/testReport/junit/UnitTests/SofaGeneralImplicitOdeSolver_test_VariationalSymplecticExplicitSolverDynamic_test/variationalSymplecticExplicitSolverDynamicTest_medium_dt_with_damping_0/)
[UnitTests.SofaGeneralImplicitOdeSolver_test/VariationalSymplecticExplicitSolverDynamic_test.variationalSymplecticExplicitSolverDynamicTest_high_dt_with_damping/0](https://ci.inria.fr/sofa-ci-dev/job/sofa-framework/job/PR-2652/CI_CONFIG=windows_vs2017,CI_PLUGINS=options,CI_TYPE=release/3/testReport/junit/UnitTests/SofaGeneralImplicitOdeSolver_test_VariationalSymplecticExplicitSolverDynamic_test/variationalSymplecticExplicitSolverDynamicTest_high_dt_with_damping_0/)
```

for a reallllly slight variation in the result (6.106226635438361e-16 diff); actually the unit tests checker is really tight 😛

**Comment by fredroy** [comment]
[ci-build][with-all-tests]

File: SofaKernel/modules/Sofa.Config/CMakeLists.txt [hunk 1]
```diff
@@ -250,8 +255,25 @@ if(MSVC)
     # Increase Number of Sections in .Obj file
     list(APPEND SOFACONFIG_COMPILE_OPTIONS "/bigobj")
     if(SOFA_VECTORIZE)
-        # SSE2 flags
-        list(APPEND SOFACONFIG_COMPILE_OPTIONS "/arch:SSE2;/fp:fast")
+        message("SOFA_VECTORIZE cmake option is deprecated and has been split between into SOFA_ENABLE_SIMD and SOFA_ENABLE_FAST_MATH\n. These 2 options will be set.")
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Config/CMakeLists.txt]:
```suggestion
        message("SOFA_VECTORIZE option is deprecated and has been split between SOFA_ENABLE_SIMD and SOFA_ENABLE_FAST_MATH. These 2 options will be enabled.")
```

File: SofaKernel/modules/Sofa.Config/CMakeLists.txt [hunk 2]
```diff
@@ -250,8 +255,25 @@ if(MSVC)
...
+        set(SOFA_ENABLE_FAST_MATH ON)
+    endif()
+    # SIMD flags
+    if(SOFA_ENABLE_SIMD)
+        include(find_avx)
+        MSVC_CHECK_FOR_AVX() # search for availability of AVX2 (otherwise set to AVX)
+        if (HAVE_AVX2_EXTENSIONS)
+            message("Your CPU supports AVX2 SIMD, set ${AVX_FLAGS} flags.")
+        elseif(HAVE_AVX_EXTENSIONS)
+            message("Your CPU supports only AVX SIMD, set ${AVX_FLAGS} flags.")
```
Comment (guparan) [inline SofaKernel/modules/Sofa.Config/CMakeLists.txt]:
```suggestion
            message("Your CPU supports AVX2 SIMD. Using flags: ${AVX_FLAGS}")
        elseif(HAVE_AVX_EXTENSIONS)
            message("Your CPU supports only AVX SIMD. Using flags: ${AVX_FLAGS}")
```

**Review by guparan — CHANGES_REQUESTED** [review]
LGTM
A few suggestions on message strings though (you know me) :angel:

File: SofaKernel/modules/Sofa.Config/CMakeLists.txt [hunk 3]
```diff
@@ -60,7 +60,9 @@ option(SOFA_WITH_OPENGL "Compile Sofa with OpenGL support. This will define the
 option(SOFA_NO_UPDATE_BBOX "Compile Sofa with the SOFA_NO_UPDATE_BBOX macro defined." OFF)
 option(SOFA_DUMP_VISITOR_INFO "Compile Sofa with the SOFA_DUMP_VISITOR_INFO macro defined." OFF)
 if(MSVC)
-    option(SOFA_VECTORIZE "Enable the use of SSE2 instructions by the compiler (Only available for MSVC)." OFF)
+    option(SOFA_VECTORIZE "(deprecated) Enable the use of SSE2 instructions by the compiler (Only available for MSVC)." OFF)
+    option(SOFA_ENABLE_SIMD "Enable the use of SIMD instructions by the compiler (AVX/AVX2 for msvc, march=native for gcc/clang)." OFF)
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Config/CMakeLists.txt:64]:
```suggestion
    option(SOFA_ENABLE_SIMD "Enable the use of SIMD instructions by the compiler (AVX/AVX2 for msvc)." OFF)
```
`SOFA_ENABLE_SIMD` is available for MSVC only, right?

File: SofaKernel/modules/Sofa.Config/CMakeLists.txt [hunk 4]
```diff
@@ -60,7 +60,9 @@ option(SOFA_WITH_OPENGL "Compile Sofa with OpenGL support. This will define the
...
+    option(SOFA_ENABLE_FAST_MATH "Enable floating-point model to fast (theorically faster but can bring unexpected results/bugs)." OFF)
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Config/CMakeLists.txt:65]:
```suggestion
    option(SOFA_ENABLE_FAST_MATH "Enable floating-point model to fast (theoretically faster but can bring unexpected results/bugs)." OFF)
```

File: SofaKernel/modules/Sofa.Config/CMakeLists.txt [hunk 5]
```diff
@@ -203,7 +205,10 @@ if(WIN32)
             # remove lots of warnings (the Wall of "normal" clang seems different of Wall of clang-cl)
             list(APPEND SOFACONFIG_COMPILE_OPTIONS -Wno-c++98-compat -Wno-c++98-compat-pedantic -Wno-double-promotion -Wno-old-style-cast -Wno-reserved-id-macro -Wno-language-extension-token -Wno-dllexport-explicit-instantiation-decl -Wno-nonportable-system-include-path -Wno-zero-as-null-pointer-constant -Wno-documentation)
             # optimization flags (not sure if necessary..)
-            list(APPEND SOFACONFIG_COMPILE_OPTIONS_RELEASE /fp:fast -march=native)
+            list(APPEND SOFACONFIG_COMPILE_OPTIONS_RELEASE -march=native)
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Config/CMakeLists.txt:208]: `march=native` means it is not possible to package SOFA for another processor than the one used to compile it.
I know Clang compilation on Windows is experimental. I just wanted to clarify it.

**Review by alxbilger — CHANGES_REQUESTED** [review]

File: SofaKernel/modules/Sofa.Config/CMakeLists.txt (hunk 4 above)
Comment (fredroy) [reply SofaKernel/modules/Sofa.Config/CMakeLists.txt:65]: Ha, I searched for that one : https://en.wiktionary.org/wiki/theorically 🤠

File: SofaKernel/modules/Sofa.Config/CMakeLists.txt (hunk 5 above)
Comment (fredroy) [reply SofaKernel/modules/Sofa.Config/CMakeLists.txt:208]: I agree, we should put some warnings  here  and there.
But actually it is the same with the compilation with AVX2 on msvc, it wont work on systems without.
So those options should not be enabled when shipping the binaries.

File: SofaKernel/modules/Sofa.Config/CMakeLists.txt (hunk 3 above)
Comment (fredroy) [reply SofaKernel/modules/Sofa.Config/CMakeLists.txt:64]: I have the same for linux/gcc/clang in an other branch (and mac/clang as well)

**Comment by alxbilger** [comment]
@fredroy Do you mind reflecting the changes from this PR in https://www.sofa-framework.org/community/doc/getting-started/build/build-options/ ?

**Pull request merged**